updated conflict resolution documents for sgw (#922)
couchbaselabs/docs-sync-gateway: modules/manage/pages/revisions.adoc

Reviewer: ```suggestion
_Documents_ and _collections_ are the basic units of data within Couchbase.
```

Since Sync Gateway 3.1 and Couchbase Lite 3.1, all data is in collections, which are part of a scope, which are part of a bucket.

All data is in a collection now.

:page-toclevels: 2@
:page-role:
:page-content: conceptual

> :description: About Sync Gateway's revision tracking using version vectors and revision caches in 4.x.

Reviewer: ```suggestion
:description: About Sync Gateway's revision tracking and caching.
```

I don't know that you need to specify 4.x, this seems confusing to me, because this documentation is already pinned to a version of Sync Gateway. This should be at the discretion of the docs team.

If you say 4.x, I'd probably say Sync Gateway 4.0+

Binary data such as images, audio and other multimedia objects are stored separately from the document in an entity known as a _blob_ (or _attachment_).

An application records each change to a document as a revision.
Document creation and deletion also generate revisions.

> Changes to blobs do not generate revisions.

Reviewer: ```suggestion
```

Blobs have always changed revisions, so this isn't true.

This was never true, and was existing bug in the documentation.

Author: Thanks for the update.

[#lbl-revs]

> === Revision Tracking in 4.x

Reviewer: See comment above whether to say Sync Gateway 4.0+ to Sync Gateway 4.0 or just rename to Revision Tracking

> Starting with Sync Gateway 4.x, revision tracking changed from revision trees to xref:server-compatibility:server-compatibility-xdcr-mobile.adoc#version-vectors[version vectors] for improved distributed system support.

Reviewer: ```suggestion
Starting with Sync Gateway 4.0, revision tracking changed from revision trees to xref:server-compatibility:server-compatibility-xdcr-mobile.adoc#version-vectors[version vectors] for improved distributed system support.
```

==== Version Vectors (4.x+)

> Sync Gateway 4.x uses xref:server-compatibility:server-compatibility-xdcr-mobile.adoc#version-vectors[version vectors] with Hybrid Logical Vectors (HLV) to track document revisions.

Reviewer: Remove Hybrid Logical Vector?

In this system:

* 1 revision exists per client that wrote a document or resolved a conflict
* Each client maintains its own component in the version vector
* The number of revisions kept is at least 5
* Only versions older than the `versionPruningWindowHrs` XDCR bucket setting are pruned

> * Version vectors enable consistent conflict resolution across distributed deployments

Reviewer: All this information is correct, as it came from me, but I think I want to frame this differently.

You probably shouldn't use this exact copy, but I think you want to use the gist of this text in place of what I wrote before.

```
A version vector stores the version history of a document across multiple sources to enable consistent conflict resolution across a distributed deployment.

Sync Gateway uses the Hybrid Logical Vector implementation  described in https://docs.couchbase.com/server/current/learn/clusters-and-availability/xdcr-enable-crossclusterversioning.html#hlv-data-maintained-in-xattr for XDCR compatability. Each element of a version vector that composed of two parts: a timestamp and a unique ID of the device [tor note: device is Couchbase Lite client or a Couchbase Server cluster, device is not the right word here.].

[Tor: from https://docs.couchbase.com/server/current/learn/clusters-and-availability/xdcr-conflict-resolution.html#time-synchronization, maybe link here and avoid this text, or duplicating is fine] Timestamps for Sync Gateway are implemented using a _Hybrid Logical Clock_.  This is a combination of a physical and a logical clock: the physical clock is the time returned by the system, in nanoseconds; the logical clock is a counter, which is incremented when the physical clock yields a value either smaller than or equal to the currently stored, physical clock-value. 

Each Couchbase Lite instance will run its own _Hybrid Logical Clock_.

* A version vector will contain 1 entry for each unique source [tor note: device/server/] consisting of the latest modified timestamp.
* The number of sources preserved is at least 5, and revisions older than https://docs.couchbase.com/server/current/learn/clusters-and-availability/xdcr-enable-crossclusterversioning.html#version-pruning-window-hrs are be pruned on document modification. [tor: use link]

The contents of version vector are implementation dependent.
Do not base any processing logic on their contents.
```

Author: Thank you for this, updated to reflect this.

==== Revision Trees (Legacy Compatibility)

> Sync Gateway 4.x maintains parallel revision trees for backward compatibility with Couchbase Lite versions earlier than 4.x.

Reviewer: see 4.0 vs 4.x discussion

These revision trees:

* Write in parallel alongside xref:server-compatibility:server-compatibility-xdcr-mobile.adoc#version-vectors[version vectors] for legacy client support
* Exist solely for backward compatibility - not used for conflict resolution in 4.x environments

> * Follow the traditional structure with generation IDs and content hashes

Reviewer: Is it reasonable to link to Sync Gateway 3.3 documentation here? not sure, but that might be the best.

* Are automatically pruned to maintain performance
* Preserve existing documents that used revision trees before the 4.x upgrade

=== Automatic Conflict Resolution

> Sync Gateway 4.x resolves conflicts automatically using Last Write Wins strategy.

Reviewer: See 4.x vs 4.0

From https://docs.couchbase.com/server/current/learn/clusters-and-availability/xdcr-conflict-resolution.html#timestamp-based-conflict-resolution I really like this text.

```suggestion
Sync Gateway 4.x resolves conflicts automatically using Last Write Wins strategy. The document timestamps  to resolve conflicts. The timestamps associated with the most recent updates of source and target documents are compared. The document whose update has the more recent timestamp prevails.
```

Author: Updated to reflect this.

[#lbl-alg]
=== Version Vector Pruning (4.x+)

Sync Gateway 4.x uses a simplified pruning approach for xref:server-compatibility:server-compatibility-xdcr-mobile.adoc#version-vectors[version vectors]:

* Sync Gateway keeps at least 5 revisions per document
* Sync Gateway prunes only versions older than the `versionPruningWindowHrs` XDCR bucket setting
* 1 revision exists per client that wrote a document or resolved a conflict

> * Pruning runs automatically when new revisions are added

Reviewer: I change this text above, I would simplify this to 

```
* The number of sources preserved is at least 5, and revisions older than https://docs.couchbase.com/server/current/learn/clusters-and-availability/xdcr-enable-crossclusterversioning.html#version-pruning-window-hrs are be pruned on document modification. [tor: use link]
```

Pruning is so niche I'd probably only mention it here.

Author: Updated to reflect this

This approach eliminates the complexity of revision tree management while ensuring sufficient history for conflict resolution.

[#lbl-rtctrl]
=== Controls

You can configure revision retention using the Configuration File's {configuration-schema-database--xref--revs-limit} setting.

For Sync Gateway 4.x with automatic conflict resolution enabled, low `revs_limit` values provide optimal performance and storage efficiency.
The Last Write Wins conflict resolution strategy supports values as low as 1.

=== Legacy Compatibility

For compatibility with Couchbase Lite versions earlier than 4.x, Sync Gateway maintains legacy revision tree structures.
These are automatically managed and do not require manual configuration in 4.x deployments.

=== Automatic Conflict Resolution (4.x)

Sync Gateway 4.x uses Last Write Wins conflict resolution by default, which eliminates the need for extensive revision history.

> This approach allows applications to use `revs_limit` values as low as 1 without affecting capabilities.

Reviewer: The default value is still 50 and maybe should be noted.

It's in https://preview.docs-test.couchbase.com/docs-sync-gateway-DOC-13591-update-conflict-resolution-sgw/sync-gateway/current/configuration/configuration-schema-database.html#revs_limit and you could link through  to this value.

I realise now that this value is generated from Sync Gateway and this needs to be modified https://github.com/couchbase/sync_gateway/blob/076e14542edcc9220839307ba675572138b12a41/docs/api/components/schemas.yaml#L1578

Reviewer: https://github.com/couchbase/sync_gateway/pull/7786#issuecomment-3358672658 where I'll update this, I'll ping back when the API docs can be rebuilt in this repo.

Author: I've updated it and also added the default value of 50

> For version vector-based documents in 4.x:

Reviewer: All documents are version vector based now, so I don't think this is necessary.

Author: Removed, thank you

> * Minimum 5 revisions are retained automatically

Reviewer: This statement is misleading. We never keep different copies of a document in Sync Gateway 4.0+ or anytime before where `allow_conflicts: false`. 

What we did kept was a history of revision ids.

Now we keep revision tree ids and a version vector. `revs_limit` applies to the number of rev tree ids stored and does not apply to the number elements in a version vector.

Author: Updated. Now clarifies that we keep revision tree IDs and version vector elements (not document copies), and that `revs_limit` only applies to revision tree IDs, not version vector elements.

> * Additional retention controlled by `versionPruningWindowHrs` XDCR bucket setting

Reviewer: It is _so unlikely_ that this would ever be configured ,we shouldn't list this more than once as to not overemphasize it.

Author: I think the older comment updated this already.

* No manual conflict resolution configuration required

NOTE: Use default `revs_limit` values for new applications.

> Higher values provide no performance benefit and can increase storage overhead.

Reviewer: This I might drop, corresponding with Sync Gateway team.

Author: Dropped this.

[#lbl-alg]
=== Version Vector Pruning (4.0+)

Sync Gateway 4.0+ uses a simplified pruning approach for xref:server-compatibility:server-compatibility-xdcr-mobile.adoc#version-vectors[version vectors]:

* The number of sources preserved is at least 5, and revisions older than xref:server:learn:clusters-and-availability/xdcr-enable-crossclusterversioning.adoc#version-pruning-window-hrs[versionPruningWindowHrs] are pruned on document modification
* 1 revision exists per client that wrote a document or resolved a conflict

This approach eliminates the complexity of revision tree management while ensuring sufficient history for conflict resolution.

[#lbl-rtctrl]
=== Controls

> You can configure revision retention using the Configuration File's {configuration-schema-database--xref--revs-limit} setting.
> 
> For Sync Gateway 4.0+ with automatic conflict resolution enabled, low `revs_limit` values provide optimal performance and storage efficiency.
> The Last Write Wins conflict resolution strategy supports values as low as 1.

Reviewer: ```suggestion
You can configure revision tree retention using the Configuration File's {configuration-schema-database--xref--revs-limit} setting. This only affects the number previous revision IDs stored in the metadata.
```

I think the language around tuning this value came from an overcorrect from `allow_conflicts=true` and it doesn't apply. This value should really never be changed.

=== Legacy Compatibility

For compatibility with Couchbase Lite versions earlier than 4.0+, Sync Gateway maintains legacy revision tree structures.
These are automatically managed and do not require manual configuration in 4.0+ deployments.

=== Automatic Conflict Resolution (4.0+)

Sync Gateway 4.0+ uses Last Write Wins conflict resolution by default, which eliminates the need for extensive revision history.

> This approach allows applications to use {configuration-schema-database--xref--revs-limit} values as low as 1 without affecting capabilities, though the default value remains 50.

Reviewer: ```suggestion
```

Drop reference to revs_limit, we actually don't want folks messing with it.
